core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -70,23 +72,174 @@ object H2OSchemaUtils {
+        val secondMap = convertToMap(second)
+        val pathItemOrdering = new Ordering[Any] {
+          override def compare(x: Any, y: Any): Int = (x, y) match {
+            case (a: Int, b: Int) => a.compareTo(b)
+            case (a: String, b: String) => a.compareTo(b)
+            case (a, b) => a.toString.compareTo(b.toString)
+          }
+        }
+        val keys = (firstMap.keySet ++ secondMap.keySet).toSeq.sorted(Ordering.Iterable(pathItemOrdering))
+        keys.map { key =>
+          (firstMap.get(key), secondMap.get(key)) match {
+            case (None, Some(StructField(name, dataType, _, _))) =>
+              FieldWithOrder(StructField(name, dataType, true), key)
+            case (Some(StructField(name, dataType, _, _)), None) =>
+              FieldWithOrder(StructField(name, dataType, true), key)
+            case (Some(StructField(name, dataType, nullable1, _)), Some(StructField(_, _, nullable2, _))) =>
+              FieldWithOrder(StructField(name, dataType, nullable1 || nullable2), key)
+          }
+        }
+      }
+      StructType(fields.map(_.field))
+  }
+
+  private def getQualifiedName(prefix: Option[String], name: String) = prefix match {
+    case None => name
+    case Some(p) => s"${p}_$name"
+  }
+
+  private def flattenField(
+      originalField: StructField,
+      row: Row,
+      index: Int,
+      path: List[Any],
+      prefix: Option[String] = None,
+      isParentNullable: Boolean = false): Seq[FieldWithOrder] = {
+    val StructField(name, dataType, nullable, _) = originalField
+    val qualifiedName = getQualifiedName(prefix, name)
+    val nullableField = isParentNullable || nullable
+    if(!row.isNullAt(index)) {
+      dataType match {
+        case BinaryType =>
+          flattenArray(row, ByteType, false, index, path, qualifiedName, nullableField)
+        case MapType(_, valueType, containsNull) =>
+          val map = row.getMap[Any, Any](index)
+          val subRow = Row.fromSeq(map.values.toSeq)
+          map.keys.zipWithIndex.flatMap { case (k, i) =>
+            val mapField = StructField(k.toString, valueType, containsNull)
+            flattenField(mapField, subRow, i, k :: path, Some(qualifiedName), nullableField)
+          }.toSeq
+        case ArrayType(elementType, containsNull) =>
+          flattenArray(row, elementType, containsNull, index, path, qualifiedName, nullableField)
+        case StructType(fields) =>
+          val subRow = row.getStruct(index)
+          fields.zipWithIndex.flatMap { case (f, i) => flattenField(f, subRow, i, i :: path, Some(qualifiedName), nullableField) }
+        case dt => FieldWithOrder(StructField(qualifiedName, dt, nullableField), path.reverse) :: Nil
+      }
+    } else {
+      Nil
+    }
+  }
+
+  private case class FieldWithOrder(field: StructField, order: Seq[Any])
+
+  private def flattenArray(
+      row: Row,
+      elementType: DataType,
+      containsNull: Boolean,
+      index: Int,
```

Comment:
DTTO: Would also remove the index and pass just the data this method works on instead of the full row and idx